### PR TITLE
Support tunnel routing in Multi-Pool IPAM mode

### PR DESCRIPTION
**Pull request by gandro**
This PR introduces support for tunnel routing with the Multi-Pool IPAM mode. This is achieved by removing the reliance on the tunnel map as a fallback mechanism and instead use Pod CIDR entries in IPCache that act as this fallback. The pod CIDR entries contain the node's tunnel endpoint and encryption key and have the `reserved:world` identity (i.e. the same identity as the existing catch-all fallback, which would have been hit before this PR). 

The bulk of the work in this PR was done by @pippolo84. This PR is a replacement for his PR https://github.com/cilium/cilium/pull/37146 with a few notable changes:

1. The first commit has been reworked to add unit tests and fix up some issues in the previous version (c.f. https://github.com/cilium/cilium/pull/37146#discussion_r1973593637).
2. A commit has been added adding some unit tests to the IPCache package emulating some of the expected corner cases.
3. Commit `node: Remove insertions and deletions to tunnel map` has been dropped. In other words, even though the tunnel map is no longer used in the `main` branch, we still populate it. This addresses concerns with regards to downgrades.
4. The Conformance Multi-Pool workflow has been extended to also test in tunnel mode.

This change will eventually allow us to remove the tunnel map completely (ref https://github.com/cilium/cilium/issues/20170).

### CODE REVIEW

**Review by aanm — APPROVED** [review]

File: pkg/node/manager/manager.go [hunk 1]
```diff
@@ -986,13 +1017,43 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
...
+}
+
 // removeNodeFromIPCache removes all addresses associated with oldNode from the IPCache,
 // unless they are present in the nodeIPsAdded, healthIPsAdded, ingressIPsAdded lists.
 // Removes ipset entry associated with oldNode if it is not present in ipsetEntries.
 //
 // The removal logic in this function should mirror the upsert logic in NodeUpdated.
 func (m *manager) removeNodeFromIPCache(oldNode nodeTypes.Node, resource ipcacheTypes.ResourceID,
-	ipsetEntries, nodeIPsAdded, healthIPsAdded, ingressIPsAdded []netip.Prefix,
+	ipsetEntries, nodeIPsAdded, healthIPsAdded, ingressIPsAdded, podCIDRsAdded []netip.Prefix,
```
Comment (pippolo84) [inline pkg/node/manager/manager.go:1058]: Nit: we should update the docstring mentioning `podCIDRsAdded`

**Review by pippolo84 — APPROVED** [review]
Thanks for completing (and improving) https://github.com/cilium/cilium/pull/37146 :pray: 

LGTM, just a (very minor) nit left inline :white_check_mark:

**Review by christarazi — APPROVED** [review]
🚀

**Comment by julianwiedmann** [comment]
:wave: Thank you for pushing this forward! Just some (potentially obvious) questions from the peanut gallery:
1. on initial upgrade, how do we learn about PodCIDRs for existing nodes and insert them into the ipcache? I suppose we replay `NodeUpdated()` as the agent initially syncs with k8s?
2. on initial upgrade, how do we handle the race condition where an endpoint's BPF program is re-generated (removing the tunnel-map lookup) **before** the PodCIDRs are populated into the ipcache? Is that just a dependency that's already enforced inside the agent?
3. for deleted nodes during agent restart, we rely on the [synthesized node deletion](https://github.com/cilium/cilium/pull/33278) feature to clean up the stale PodCIDRs?

**Review by ghost — APPROVED** [review]
docs good

**Review by doniacld — APPROVED** [review]
Nice 🌍

**Comment by gandro** [comment]
So I just had the realization that keeping the old tunnel map around for downgrade purposes is pretty pointless, since we are re-creating it every time agent restarts: https://github.com/cilium/cilium/blob/c6f05a5360918e646126d313a20dd9fb1b4db613/daemon/cmd/datapath.go#L162

So it might make sense to re-add `node: Remove insertions and deletions to tunnel map` and remove the tunnel map completely.

**Comment by julianwiedmann** [comment]
> > ```
> > 2. on initial upgrade, how do we handle the race condition where an endpoint's BPF program is re-generated (removing the tunnel-map lookup) **before** the PodCIDRs are populated into the ipcache? Is that just a dependency that's already enforced inside the agent?
> > ```
> 
> Yes, this the dependency is enforced by the agent. Endpoints are not regenerated until IPCache is fully populated. [We wait for the K8s caches to be synced (and the initial IPCache revision to be realized)](https://github.com/cilium/cilium/blob/49d3e565c5860ba1b4a5b18eefa1c0a29b75a939/daemon/cmd/daemon_main.go#L1668-L1689) before we allow endpoint regeneration to start.
> 
> > ```
> > 3. for deleted nodes during agent restart, we rely on the [synthesized node deletion](https://github.com/cilium/cilium/pull/33278) feature to clean up the stale PodCIDRs?
> > ```
> 
> Since IPCache is re-created when the agent restarts, there are no stale pod CIDRs to clean up (at least with regards to IPCache - when it comes to the `$podCIDR via cilium_host` routes installed by `linuxNodeHandler`, we do rely on those - but that behavior in unchanged with this PR).

Awesome, thank you! This helps a lot to understand the bigger picture. And it also clarifies what happens on downgrade - as the IPCache is rebuilt, the PodCIDR entries naturally disappear again.

**Comment by snowhanse** [comment]
Just wanted to add to this to say that this is exactly what I've been looking for. 
I've done a _much_ cruder version of this, proving out that cilium can support my use case of having default pod ip pool use tunnel and separate pool as routable with IPs advertised over BGP.

So I'll also wait for this one to merge, and in the meantime use it as a base for my testing.

**Comment by pippolo84** [comment]
Superseded by https://github.com/cilium/cilium/pull/38483
